src/ServiceInsight/MessageFlow/MessageFlowViewModel.cs

```diff
@@ -259,18 +260,39 @@ void LinkConversationNodes(IEnumerable<MessageNode> relatedMessagesTask)
         {
             foreach (var msg in relatedMessagesTask)
             {
-                if (msg.Message.RelatedToMessageId == null &&
-                    msg.Message.RelatedToMessageId != msg.Message.MessageId)
+                if (string.IsNullOrEmpty(msg.Message.RelatedToMessageId) && msg.Message.RelatedToMessageId != msg.Message.MessageId)
                 {
                     continue;
                 }
 
-                // [CM] I don't know how it's happening, but a user reported an
-                // error where multiple results were returned from this query.
                 var parentMessages = nodeMap.Values.Where(m =>
-                    m.Message != null && m.Message.ReceivingEndpoint != null && m.Message.SendingEndpoint != null &&
-                    m.Message.MessageId == msg.Message.RelatedToMessageId &&
-                    m.Message.ReceivingEndpoint.Name == msg.Message.SendingEndpoint.Name);
+                    m.Message != null && m.Message.ReceivingEndpoint != null && m.Message.SendingEndpoint != null
```

Comment:
The code potentially multi-iterates the IEnumerable. Maybe convert it to the list here? and update th usages e.g. `.Count()` to `.Count` etc.?

Reply:
Resolved via 7755458168c0a8d40dfdc8c766cb061eb0dd8cc6